### PR TITLE
Fix delete

### PR DESCRIPTION
**Pull request by enxebre**
This lets the nodePool controller to explicitly delete its owned machineSet. Removing the owner ref is not enough as it has other from the cluster resource.

It also move the default nodePool to be created by the openshiftCluster controller. Otherwise openshiftCluster deletion is not able to delete de default nodePool.

We are blocking hcp controller on https://github.com/openshift-hive/hypershift/blob/main/hypershift-operator/controllers/infra.go#L42 though the ign endpoint will only be given after deploying the service which is only applied after that https://github.com/openshift-hive/hypershift/blob/main/hypershift-operator/controllers/hosted_controlplane_controller.go#L156
This results in a deadlock loop which is broken by this PR.

Remove foreground deletion and relax timeouts for e2e

### CODE REVIEW

File: hypershift-operator/controllers/nodepool_controller.go [hunk 1]
```diff
@@ -73,17 +75,40 @@ func (r *NodePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+			return reconcile.Result{}, fmt.Errorf("failed to generate worker machineset: %w", err)
+		}
+		if err := r.Delete(ctx, machineSet); err != nil && !apierrors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("failed to delete cluster: %w", err)
+		}
+
+		if controllerutil.ContainsFinalizer(nodePool, finalizer) {
+			controllerutil.RemoveFinalizer(nodePool, finalizer)
+			if err := r.Update(ctx, nodePool); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from cluster: %w", err)
```
Comment (csrwng) [inline hypershift-operator/controllers/nodepool_controller.go]: nit: "failed to remove finalizer from nodepool"

File: hypershift-operator/controllers/nodepool_controller.go [hunk 2]
```diff
@@ -73,17 +75,40 @@ func (r *NodePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+			}
+		}
+		r.Log.Info("Delete machineSet", "machineset", machineSet.GetName())
```
Comment (csrwng) [inline hypershift-operator/controllers/nodepool_controller.go]: nit: "Deleted"

File: hypershift-operator/controllers/nodepool_controller.go [hunk 3]
```diff
@@ -73,17 +75,40 @@ func (r *NodePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
 		return ctrl.Result{}, nil
 	}
 
+	// Ensure the cluster has a finalizer for cleanup
```
Comment (csrwng) [inline hypershift-operator/controllers/nodepool_controller.go]: nit: s/cluster/nodepool/

File: hypershift-operator/controllers/nodepool_controller.go [hunk 4]
```diff
@@ -73,17 +75,40 @@ func (r *NodePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+	if !controllerutil.ContainsFinalizer(nodePool, finalizer) {
+		controllerutil.AddFinalizer(nodePool, finalizer)
+		if err := r.Update(ctx, nodePool); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer to cluster: %w", err)
```
Comment (csrwng) [inline hypershift-operator/controllers/nodepool_controller.go]: nit: s/cluster/nodepool/

File: hypershift-operator/controllers/openshiftcluster_controller.go [hunk 5]
```diff
@@ -190,6 +215,16 @@ func (r *OpenShiftClusterReconciler) delete(ctx context.Context, name, namespace
...
 	}
+	defaultNodePool := &hyperv1.NodePool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%v-default", name),
+			Namespace: namespace,
+		},
+	}
+
+	if err := r.Delete(ctx, defaultNodePool); err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed to delete cluster: %w", err)
```
Comment (csrwng) [inline hypershift-operator/controllers/openshiftcluster_controller.go:226]: nit: s/cluster/nodepool/

File: hypershift-operator/controllers/openshiftcluster_controller.go [hunk 6]
```diff
@@ -190,6 +215,16 @@ func (r *OpenShiftClusterReconciler) delete(ctx context.Context, name, namespace
 			Namespace: namespace,
 		},
 	}
+	defaultNodePool := &hyperv1.NodePool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%v-default", name),
```
Comment (csrwng) [inline hypershift-operator/controllers/openshiftcluster_controller.go]: s/%v/%s/

File: hypershift-operator/controllers/infra.go [hunk 7]
```diff
@@ -38,8 +38,7 @@ type InfrastructureStatus struct {
 func (s InfrastructureStatus) IsReady() bool {
 	return len(s.APIAddress) > 0 &&
 		len(s.OAuthAddress) > 0 &&
-		len(s.VPNAddress) > 0 &&
-		len(s.IgnitionProviderAddress) > 0
```
Comment (csrwng) [inline hypershift-operator/controllers/infra.go]: we can drop the check, but we should also get the route hostname from the spec instead of status

**Review by csrwng — CHANGES_REQUESTED** [review]
Hey @enxebre looks good. Some nits, also why do we want to remove foreground deletion in the e2e?

**Comment by enxebre** [comment]
>Hey @enxebre looks good. Some nits, also why do we want to remove foreground deletion in the e2e?

with background deletion cluster resource enforce ordering of dependants deletion programatically, first machine scalable resources, then referenced controlPlane, then referenced infrastructure resource.
I believe with foreground deletion this ordering is not respected, as soon as you delete the cluster resource the garbage collector triggers deletion of all dependents. https://kubernetes.slack.com/archives/C8TSNPY4T/p1610974730023300
https://github.com/kubernetes-sigs/cluster-api/issues/4085

**Pull request merged**